server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/SpanServiceElasticsearch.java

```diff
@@ -192,6 +202,85 @@ public void clear(String tenantId) {
         }
     }
 
+    @Override
+    public Trace getTraceFragment(String tenantId, String id) {
+        Span span = getSpan(tenantId, id);
+
+        InteractionNode interactionNode = spanToNode(span);
+        interactionNode.setNodes(recursiveTraceFragment(tenantId, id));
+
+        Trace trace = spanToTrace(span);
+        trace.getNodes().add(interactionNode);
+
+        return trace;
+    }
+
+
+    private List<Node> recursiveTraceFragment(String tenantId, String spanId) {
+        List<Span> spanChildren = getChildren(tenantId, spanId);
+        if (spanChildren == null) {
+            return new LinkedList<>();
+        }
+
+        List<Node> nodes = new ArrayList<>();
+
+        for (Span span: spanChildren) {
+            if (span.serverSpan()) {
+                continue;
+            }
+
+            InteractionNode node = spanToNode(span);
+            nodes.add(node);
+
+            node.setNodes(recursiveTraceFragment(tenantId, span.getId()));
+        }
+
+        return nodes;
+    }
+
+    private Trace spanToTrace(Span span) {
+        if (span == null) {
+            throw new NullPointerException();
+        }
+
+        List<Property> spanProperties = span.properties();
+
+        Trace trace = new Trace();
+        trace.setId(span.getId());
+        trace.setStartTime(span.getTimestamp());
+        trace.setHostAddress(Span.ipv4Address(spanProperties));
+        trace.setProperties(new HashSet<>(span.properties()));
+
+//        TODO remove
+//        trace.setBusinessTransaction()
+//        trace.setHostName()
+//        trace.setPrincipal()
+
+        return trace;
+    }
+
+    private InteractionNode spanToNode(Span span) {
+        String endpointType = "http";
+        String url = span.url() != null ? span.url().toString() : null;
+
+        InteractionNode node;
+        if (span.serverSpan()) {
+            node = new Consumer(url, endpointType);
+        }else if (span.clientSpan()) {
+            node = new Producer(url, endpointType);
+        } else {
+            node = new Component(url, null);
+        }
+
+        node.setBaseTime(TimeUnit.NANOSECONDS.convert(span.getTimestamp(), TimeUnit.MICROSECONDS));
+        /**
+         * TODO set duration ss - sr? or cr - cs which captures network latency?
+         */
+//        node.setDuration();
```

Comment:
It should be between the two annotations in the same span, so either sr -> ss, or cs -> cr.

Not sure what should happen with a component, as I don't know how they record the time taken within a span that is not a client or server?